### PR TITLE
[WIP] Changes comparison for ONNX conversion

### PR DESCRIPTION
**Pull request by ganik**

### CODE REVIEW

File: DeBERTa/deberta/disentangled_attention.py [hunk 1]
```diff
@@ -174,7 +174,9 @@ def linear(w,b,x):
         if self.talking_head:
             attention_scores = self.head_logits_proj(attention_scores.permute(0,2,3,1)).permute(0,3,1,2)
 
-        attention_probs = XSoftmax.apply(attention_scores, attention_mask, -1)
+        #attention_probs = XSoftmax.apply(attention_scores, attention_mask, -1)
+        nodex = torch.nn.Softmax(-1)
+        attention_probs = nodex(attention_scores)
```
Comment (BigBird01) [inline DeBERTa/deberta/disentangled_attention.py:179]: You need to apply the mask to the attention_scores by add it with a -inf number, e.g. -10000

Comment (ganik) [reply DeBERTa/deberta/disentangled_attention.py:179]: Something like this ? : 
attention_probs = nodex(attention_scores * attention_mask)

Comment (ganik) [reply DeBERTa/deberta/disentangled_attention.py:179]: And what about Dropout changes? How to apply mask there?

File: DeBERTa/apps/train.py [hunk 2]
```diff
@@ -217,9 +218,63 @@ def run_predict(args, model, device, eval_data, prefix=None):
...
+  train_dataloader = DataLoader(train_data, batch_sampler=batch_sampler, num_workers=args.workers, pin_memory=True)
+  torch.cuda.empty_cache()
+  for step, batch in enumerate(AsyncDataLoader(train_dataloader, 100)):
+    #import pdb
+    #pdb.set_trace()
+    batch = batch_to(batch, device)
+    with torch.no_grad():
+      trainer.train_step(batch['input_ids'], batch['type_ids'], batch['position_ids'], batch['input_mask'], batch['labels'])
+      # conversion fails now with:
+      # site-packages/torch/onnx/utils.py:617: UserWarning: ONNX export failed on ATen operator broadcast_tensors
```
Comment (ganik) [inline DeBERTa/apps/train.py:272]: Going to follow up with the team to figure out how to workaround this issue

File: DeBERTa/apps/sequence_classification.py [hunk 3]
```diff
@@ -68,4 +68,4 @@ def forward(self, input_ids, type_ids=None, input_mask=None, labels=None, positi
         label_confidence = 1
         loss = -((log_softmax(logits)*labels).sum(-1)*label_confidence).mean()
 
-    return (logits,loss)
+    return (loss, logits)
```
Comment (ganik) [inline DeBERTa/apps/sequence_classification.py:71]: During onnx conversion had error message requiring loss to be a first scalar in tuple

File: DeBERTa/deberta/ops.py [hunk 4]
```diff
@@ -115,7 +115,11 @@ def backward(ctx, grad_output):
     else:
       return grad_output, None
 
-class StableDropout(torch.nn.Module):
+class StableDropout(torch.nn.Dropout):
```
Comment (ganik) [inline DeBERTa/deberta/ops.py:118]: Replacing with nn.Dropout regresses model performance significantly, investigation needed.

File: DeBERTa/optims/fp16_optimizer.py [hunk 5]
```diff
@@ -20,6 +20,8 @@
 from ..utils import get_logger
 logger=get_logger()
 
+# Lines below tightly couple DeBerta with torch 1.3 
+# TBD refactor or port to torch 1.6
 lib = ctypes.cdll.LoadLibrary(None)
 lib.THCudaHalfTensor_normall.argtypes=[ctypes.c_void_p, ctypes.c_void_p]
 lib.THCudaHalfTensor_normall.restype = ctypes.c_float
```
Comment (ganik) [inline DeBERTa/optims/fp16_optimizer.py:27]: Refactor to get rid of torch 1.3 tight coupling

File: DeBERTa/apps/train.py (hunk 2 above)
Comment (ganik) [inline DeBERTa/apps/train.py:272]: mse_loss implementation in https://github.com/pytorch/pytorch/blob/master/torch/nn/functional.py#L2682 uses 2 ops that are not implemented: broadcast_tensors() and mse_loss(). Working around this to get unblocked, made a patch:
        #expanded_input, expanded_target = torch.broadcast_tensors(input, target)
        expanded_input = input + torch.zeros(target.size())
        expanded_target = target + torch.zeros(input.size())
        #ret = torch._C._nn.mse_loss(expanded_input, expanded_target, _Reduction.get_enum(reduction))
        t = expanded_input - expanded_target
        t = t * t
        ret = torch.mean(t)

File: DeBERTa/deberta/bert.py [hunk 6]
```diff
@@ -160,7 +160,7 @@ def get_attention_mask(self, attention_mask):
     if attention_mask.dim()<=2:
       extended_attention_mask = attention_mask.unsqueeze(1).unsqueeze(2)
       attention_mask = extended_attention_mask*extended_attention_mask.squeeze(-2).unsqueeze(-1)
-      attention_mask = attention_mask.byte()
+      attention_mask = attention_mask.int()
```
Comment (ganik) [inline DeBERTa/deberta/bert.py:163]: This to fix the below error:
onnxruntime.capi.onnxruntime_pybind11_state.InvalidGraph: [ONNXRuntimeError] : 10 : INVALID_GRAPH : This is an invalid model. Type Error: Type 'tensor(uint8)' of input parameter (234) of operator (Sub) in node (Sub_3360) is invalid.

File: DeBERTa/deberta/disentangled_attention.py [hunk 7]
```diff
@@ -174,7 +177,9 @@ def linear(w,b,x):
         if self.talking_head:
             attention_scores = self.head_logits_proj(attention_scores.permute(0,2,3,1)).permute(0,3,1,2)
 
-        attention_probs = XSoftmax.apply(attention_scores, attention_mask, -1)
+        #attention_probs = XSoftmax.apply(attention_scores, attention_mask, -1)
+        nodex = torch.nn.Softmax(-1)
```
Comment (ganik) [inline DeBERTa/deberta/disentangled_attention.py:181]: Replacing XSoftmax with Softmax

File: DeBERTa/deberta/disentangled_attention.py [hunk 8]
```diff
@@ -77,8 +77,9 @@ def __init__(self, config):
         self.attention_head_size = int(config.hidden_size / config.num_attention_heads)
         self.all_head_size = self.num_attention_heads * self.attention_head_size
         self.in_proj = torch.nn.Linear(config.hidden_size, self.all_head_size*3, bias=False)
-        self.q_bias = torch.nn.Parameter(torch.zeros((self.all_head_size), dtype=torch.float))
-        self.v_bias = torch.nn.Parameter(torch.zeros((self.all_head_size), dtype=torch.float))
+        # Looks like params below are never updated and const, so removing them
+        #self.q_bias = torch.nn.Parameter(torch.zeros((self.all_head_size), dtype=torch.float))
```
Comment (ganik) [inline DeBERTa/deberta/disentangled_attention.py:81]: q_bias and v_bias are always const, so commenting them out

File: DeBERTa/deberta/ops.py (hunk 4 above)
Comment (ganik) [reply DeBERTa/deberta/ops.py]: fixed

**Comment by ganik** [comment]
> Instead of change code every where, why not just change StableDropout?

Previous iterations i tried to redefine StableDropout to inherit from nn.Dropout, but it led to regression in model stats. Could not figure out why. If i do change this way there is no regression. Something was missing with just redefining StableDropout.
